src/librustc_typeck/check/mod.rs

```diff
@@ -3881,7 +3881,15 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             hir::StmtKind::Item(_) => {}
             hir::StmtKind::Expr(ref expr) => {
                 // Check with expected type of `()`.
-                self.check_expr_has_type_or_error(&expr, self.tcx.mk_unit());
+
+                self.check_expr_has_type_or_error(&expr, self.tcx.mk_unit(), |err| {
+                    err.span_suggestion_short(
+                        expr.span.shrink_to_hi(),
+                        "consider using a semicolon here",
```

Comment:
Would be good to refactor the suggestion into a method since you've used it twice in this PR. I have a feeling it's used elsewhere also.